test/extensions/config_subscription/grpc/xds_failover_integration_test.cc

```diff
@@ -0,0 +1,575 @@
+  test_server_->waitForGaugeEq("cluster.cluster_0.warming_state", 1);
+
+  // Envoy has received a CDS response, it means the primary is available.
+  // Now disconnect the primary.
+  xds_stream_->finishGrpcStream(Grpc::Status::Internal);
+
+  // In this case (received a response), both EnvoyGrpc and GoogleGrpc keep the connection open.
+  result = xds_connection_->waitForNewStream(*dispatcher_, xds_stream_);
+  RELEASE_ASSERT(result, result.message());
+  // Immediately fail the connection.
+  xds_stream_->finishGrpcStream(Grpc::Status::Internal);
+
+  // Ensure that Envoy still attempts to connect to the primary,
+  // and keep disconnecting a few times and validate that the failover
+  // connection isn't attempted.
+  for (int i = 3; i < 5; ++i) {
+    // EnvoyGrpc will disconnect if the gRPC stream is immediately closed (as
+    // done above).
+    if (clientType() == Grpc::ClientType::EnvoyGrpc) {
+      result = xds_upstream_->waitForHttpConnection(*dispatcher_, xds_connection_);
+      RELEASE_ASSERT(result, result.message());
+    }
+    result = xds_connection_->waitForNewStream(*dispatcher_, xds_stream_);
+    RELEASE_ASSERT(result, result.message());
+    // Immediately fail the connection.
+    xds_stream_->finishGrpcStream(Grpc::Status::Internal);
+  }
+
+  // When GoogleGrpc is used, a connection to the failover_xds_upstream will be
+  // attempted, but no stream will be created. When EnvoyGrpc is used, no
+  // connection to the failover will be attempted.
+  if (clientType() == Grpc::ClientType::EnvoyGrpc) {
+    // A failover connection should not be attempted, so a failure is expected here.
+    // Setting smaller timeout to avoid long execution times.
+    EXPECT_FALSE(failover_xds_upstream_->waitForHttpConnection(
+        *dispatcher_, failover_xds_connection_, std::chrono::seconds(1)));
+  } else {
+    result = failover_xds_upstream_->waitForHttpConnection(*dispatcher_, failover_xds_connection_);
+    RELEASE_ASSERT(result, result.message());
+    EXPECT_FALSE(failover_xds_connection_->waitForNewStream(*dispatcher_, failover_xds_stream_,
+                                                            std::chrono::seconds(1)));
+  }
+
+  // Allow a connection to the primary.
+  // Expect a connection to the primary when using EnvoyGrpc.
+  // In case GoogleGrpc is used the current connection will be reused (new stream).
+  if (clientType() == Grpc::ClientType::EnvoyGrpc) {
+    result = xds_upstream_->waitForHttpConnection(*dispatcher_, xds_connection_);
+    RELEASE_ASSERT(result, result.message());
+  }
+  result = xds_connection_->waitForNewStream(*dispatcher_, xds_stream_);
+  xds_stream_->startGrpcStream();
+
+  // The rest will be a normal primary source xDS back and forth.
+  validateAllXdsResponsesAndDataplaneRequest(xds_stream_.get());
+}
+
+// Validate that once failover answers, primary will not be used, even after disconnecting.
+TEST_P(XdsFailoverAdsIntegrationTest, NoPrimaryUseAfterFailoverResponse) {
```

Comment:
what would be happening if primary is not responding and failover as well, would it try to again connect to primary?

Reply:
Envoy will alternate between the primary and failover over and over until one of them respondes. We've got a unit test that covers this in https://github.com/envoyproxy/envoy/blob/46f86da8a471e6a23ce5cf1431cc9abce2dc260e/test/extensions/config_subscription/grpc/grpc_mux_failover_test.cc#L344

Reply:
great, just wanted make sure this is covered by tests.